File: project/src/com/LangStackTest/NormalTypeClass.java
package com.LangStackTest;

import java.util.ArrayList;

public class NormalTypeClass {
    private int                      mInt;
    private long                     mLonglong;
    private double                   mDouble;
    private String                   mString;
    private EasyTypeClass            mClass;
    private EasyTypeClass            mClassNULL;

    private ArrayList<Integer>       mIntList;
    private ArrayList<Long>          mLonglongList;
    private ArrayList<Double>        mDoubleList;
    private ArrayList<String>        mStringList;
    private ArrayList<EasyTypeClass> mClassList;

    private ArrayList<Integer>       mIntListEmpty;
    private ArrayList<Long>          mLongListEmpty;
    private ArrayList<Long>          mLonglongListEmpty;
    private ArrayList<Double>        mDoubleListEmpty;
    private ArrayList<String>        mStringListEmpty;
    private ArrayList<EasyTypeClass> mClassListEmpty;

    public NormalTypeClass()
    {
        mInt = 0;
        mLonglong = 0;
        mDouble = 0;
        mString = new String();
        mClass = null;
        mClassNULL = null;
        mIntList = new ArrayList<Integer>();
        mLonglongList = new ArrayList<Long>();
        mDoubleList = new ArrayList<Double>();
        mStringList = new ArrayList<String>();
        mClassList = new ArrayList<EasyTypeClass>();
        mIntListEmpty = new ArrayList<Integer>();
        mLongListEmpty = new ArrayList<Long>();
        mLonglongListEmpty = new ArrayList<Long>();
        mDoubleListEmpty = new ArrayList<Double>();
        mStringListEmpty = new ArrayList<String>();
        mClassListEmpty = new ArrayList<EasyTypeClass>();
    }

    public NormalTypeClass(int seed)
    {
        mInt = seed;
        mLonglong = seed * 10001000L;
        mDouble = seed * 0.001001;
        mClass = null;
        mClassNULL = null;
        mIntList = new ArrayList<Integer>();
        mLonglongList = new ArrayList<Long>();
        mDoubleList = new ArrayList<Double>();
        mStringList = new ArrayList<String>();
        mClassList = new ArrayList<EasyTypeClass>();
        mIntListEmpty = new ArrayList<Integer>();
        mLongListEmpty = new ArrayList<Long>();
        mLonglongListEmpty = new ArrayList<Long>();
        mDoubleListEmpty = new ArrayList<Double>();
        mStringListEmpty = new ArrayList<String>();
        mClassListEmpty = new ArrayList<EasyTypeClass>();

        mString = new String();
        mString = String.format("%d-%d", seed, seed);
        EasyTypeClass temp = new EasyTypeClass();
        temp.mDouble = mDouble;
        temp.mInt = mInt;
        temp.mLonglong = mLonglong;
        temp.mSpecial = mString + ":::%0%1%;;;";
        temp.mNormal = mString;
        mClass = temp;

        for (int i = 1; i < 4; ++i)
        {
            mIntList.add(mInt + i);
            mLonglongList.add(mLonglong + i);
            mDoubleList.add(mDouble + i);
            mStringList.add(mString);

            temp = new EasyTypeClass();
            temp.mDouble = mDouble;
            temp.mInt = mInt + i;
            temp.mLonglong = mLonglong;
            temp.mSpecial = mString + ":::%%%;;;";
            temp.mNormal = mString;
            mClassList.add(temp);
        }
    }

    public boolean equals(NormalTypeClass rhs)
    {
		boolean ret = mInt == rhs.mInt && mLonglong == rhs.mLonglong
				&& mDouble == rhs.mDouble && mString.equals(rhs.mString)
				&& mClass.equals(rhs.mClass) && mClassNULL == rhs.mClassNULL;

        ret &= equals(mIntList, rhs.mIntList);
        ret &= equals(mDoubleList, rhs.mDoubleList);
        ret &= equals(mStringList, rhs.mStringList);
        ret &= equals(mClassList, rhs.mClassList);
        ret &= equals(mIntListEmpty, rhs.mIntListEmpty);
        ret &= equals(mLongListEmpty, rhs.mLongListEmpty);
        ret &= equals(mLonglongListEmpty, rhs.mLonglongListEmpty);
        ret &= equals(mDoubleListEmpty, rhs.mDoubleListEmpty);
        ret &= equals(mStringListEmpty, rhs.mStringListEmpty);
        ret &= equals(mClassListEmpty, rhs.mClassListEmpty);

        return ret;
    }

    private boolean equals(ArrayList<?> lhs, ArrayList<?> rhs)
    {
        if (lhs.size() != rhs.size())
        {
            return false;
        }

        for (int i = 0; i < lhs.size(); ++i)
        {
            if (!lhs.get(i).equals(rhs.get(i)))
            {
                return false;
            }
        }

        return true;
    }
}

File: cpptest/src/NormalTypeClass.cpp
/*************************************************************************
 ** 版权保留(C), 2001-2014, 浙江大华技术股份有限公司.
 ** 版权所有.
 **
 ** $Id$
 **
 ** 功能描述   :
 **
 ** 修改历史     : 2015年1月13日 zhu_long Modification
*************************************************************************/

#include "NormalTypeClass.h"

#include <iostream>
#include "ICallback.h"

CNormalTypeClass::CNormalTypeClass() : CNormalTypeClass(5) {}

CNormalTypeClass::CNormalTypeClass(int seed) : m_int(seed),
        m_longlong(seed * 10001000LL), m_double(seed * 0.001001),
        m_class(0), m_classNULL(0), m_handler(0)
{
    char buf[256];
    sprintf(buf, "%d-%d", seed, seed);
    m_string  = buf;
    CEasyTypeClass* temp = new CEasyTypeClass();
    temp->m_double      = m_double;
    temp->m_int         = m_int;
    temp->m_longlong    = m_longlong;
    temp->m_special     = m_string + ":::%0%1%;;;";
    temp->m_normal      = m_string;
    m_class             = temp;

    for (int i = 1; i < 4; ++i)
    {
        m_intList.push_back(m_int + i);
        m_longlongList.push_back(m_longlong + i);
        m_doubleList.push_back(m_double + i);
        m_stringList.push_back(m_string);

        temp = new CEasyTypeClass();
        temp->m_double   = m_double;
        temp->m_int      = m_int + i;
        temp->m_longlong = m_longlong;
        temp->m_special  = m_string + ":::%%%;;;";
        temp->m_normal   = m_string;
        m_classList.push_back(temp);
    }
}

CNormalTypeClass::~CNormalTypeClass()
{
    if (m_class)
    {
        delete m_class;
        m_class = NULL;
    }

    for (auto it = m_classList.begin(); it != m_classList.end(); ++it)
    {
        delete *it;
    }
    m_classList.clear();
}

void CNormalTypeClass::dump()
{
    std::cout << "m_int:\t\t\t" << m_int << std::endl;
    std::cout << "m_longlong:\t\t" << m_longlong << std::endl;
    std::cout << "m_double:\t\t" << m_double << std::endl;
    std::cout << "m_string:\t\t" << m_string << std::endl;

    std::cout << "CEasyTypeClass:\t\t" << std::endl;
    dynamic_cast<CEasyTypeClass*>(m_class)->dump();

    for (auto it = m_classList.begin(); it != m_classList.end(); ++it)
    {
        std::cout << "CEasyTypeClass:\t\t" << std::endl;
        dynamic_cast<CEasyTypeClass*>(*it)->dump();
    }

    if (0 != m_handler)
    {
        ICallBack *callback = (ICallBack *)m_handler;
        callback->interface1();
        std::cout << __FILE__ << "-" << __LINE__ << ":" << callback->interface2(33) << std::endl;
        std::cout << __FILE__ << "-" << __LINE__ << ":" << callback->interface3({1,2}, nullptr) << std::endl;
        std::cout << __FILE__ << "-" << __LINE__ << ":" << callback->interface4() << std::endl;

        std::cout << __FILE__ << "-" << __LINE__ << ":" <<
                callback->interface3({1,2}, new CEasyTypeClass()) << std::endl;
    }
}

double CNormalTypeClass::test(int i, long long j, std::list<int> ilist, std::list<long long> llist)
{
    std::cout << "i:\t\t\t" << i << std::endl;
    std::cout << "j:\t\t\t" << j << std::endl;

    std::cout << "m_int:\t\t\t" << m_int << std::endl;
    std::cout << "m_longlong:\t\t" << m_longlong << std::endl;
    std::cout << "m_double:\t\t" << m_double << std::endl;
    std::cout << "m_string:\t\t" << m_string << std::endl;

    for (auto &ii : ilist) std::cout << "ii:\t\t\t\t" << ii << std::endl;
    for (auto &ll : llist) std::cout << "ll:\t\t\t\t" << ll << std::endl;

    return 0.01;
}

void CNormalTypeClass::attach(long long handler)
{
    m_handler = handler;
}

void CNormalTypeClass::detach(long long handler)
{
    if (m_handler == handler)
    {
        m_handler = 0;
    }
}

bool CNormalTypeClass::operator==(const CNormalTypeClass&rhs) const
{
    bool ret = isEqual(m_int,               rhs.m_int);
    ret     &= isEqual(m_longlong,          rhs.m_longlong);
    ret     &= isEqual(m_double,            rhs.m_double);
    ret     &= isEqual(m_string,            rhs.m_string);
    ret     &= isEqual(m_class,             rhs.m_class);
    ret     &= isEqual(m_classNULL,         rhs.m_classNULL);
    ret     &= isEqual(m_intList,           rhs.m_intList);
    ret     &= isEqual(m_longlongList,      rhs.m_longlongList);
    ret     &= isEqual(m_doubleList,        rhs.m_doubleList);
    ret     &= isEqual(m_stringList,        rhs.m_stringList);
    ret     &= isEqual(m_classList,         rhs.m_classList);
    ret     &= isEqual(m_intListEmpty,      rhs.m_intListEmpty);
    ret     &= isEqual(m_longlongListEmpty, rhs.m_longlongListEmpty);
    ret     &= isEqual(m_doubleListEmpty,   rhs.m_doubleListEmpty);
    ret     &= isEqual(m_stringListEmpty,   rhs.m_stringListEmpty);
    ret     &= isEqual(m_classListEmpty,    rhs.m_classListEmpty);

    return ret;
}
